pkg/image/controller/controller.go

```diff
@@ -22,6 +31,91 @@ type ImportController struct {
 	mappings client.ImageStreamMappingsNamespacer
 	// injected for testing
 	client dockerregistry.Client
+
+	stopChan chan struct{}
+
+	imageStreamController *framework.Controller
+
+	work           chan *api.ImageStream
+	workingSet     sets.String
+	workingSetLock sync.Mutex
+
+	// this should not be larger the capacity of the work channel
+	numParallelImports int
+}
+
+func NewImportController(isNamespacer client.ImageStreamsNamespacer, ismNamespacer client.ImageStreamMappingsNamespacer, parallelImports int, resyncInterval time.Duration) *ImportController {
+	c := &ImportController{
+		streams:  isNamespacer,
+		mappings: ismNamespacer,
+
+		numParallelImports: parallelImports,
+		work:               make(chan *api.ImageStream, 20*parallelImports),
+		workingSet:         sets.String{},
+	}
+
+	_, c.imageStreamController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return c.streams.ImageStreams(kapi.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(resourceVersion string) (watch.Interface, error) {
+				return c.streams.ImageStreams(kapi.NamespaceAll).Watch(labels.Everything(), fields.Everything(), resourceVersion)
+			},
+		},
+		&api.ImageStream{},
+		resyncInterval,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    c.imageStreamAdded,
+			UpdateFunc: c.imageStreamUpdated,
+		},
+	)
+
+	return c
+}
+
+// Runs controller loops and returns immediately
+func (c *ImportController) Run() {
+	if c.stopChan == nil {
+		c.stopChan = make(chan struct{})
+		go c.imageStreamController.Run(c.stopChan)
+
+		for i := 0; i < c.numParallelImports; i++ {
+			go c.handleImport()
```

Comment:
> One other thing: instead of having both Run and Stop you could just wait on stopChan after spawning the workers, and close afterwards. Similarly to how it's done in job's controller. This should simplify the code a bit, imho.

I personally prefer the structured `Run()`, `Stop()` pair.  It seems like the total work is the same, the separate channel simply means that something else is tracking your state for you, allowing the controller to leak its mechanics to calling code.  The only power gained is the ability to control multiple controllers with a single channel, but that seems to be of dubious value.